### PR TITLE
Adding code for training your own WaveNet that is ready to work with NV-WaveNet inference

### PR DESCRIPTION
**Pull request by RPrenger**

### CODE REVIEW

**Comment by reidsanders** [comment]
This worked without apparent issue for me (using defaults). Thanks!

File: pytorch/wavenet.py [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+    def get_cond_input(self, features):
+        """
+        Takes in features and gets the 2*R x batch x # layers x samples tensor
+        """
+        # TODO(rcosta): trim conv artifacts. mauybe pad spec to kernel multiple
+        cond_input = self.upsample(features)
+        time_cutoff = self.upsample.kernel_size[0] - self.upsample.stride[0]
+        cond_input = cond_input[:, :, :-time_cutoff]
+        cond_acts = [layer.conv(cond_input).data for layer in self.cond_layers]
```
Comment (PetrochukM) [inline pytorch/wavenet.py]: Use a 2d Conv instead, for performance.

File: pytorch/wavenet.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+                                    bias=bias)
+
+        torch.nn.init.xavier_uniform(
+            self.conv.weight, gain=torch.nn.init.calculate_gain(w_init_gain))
+
+    def forward(self, signal):
+        conv_signal = self.conv(signal)
+        if self.is_causal:
+            # Removing padding by-products at the end
+            conv_signal = conv_signal[:, :, :signal.size(2)]
```
Comment (PetrochukM) [inline pytorch/wavenet.py]: Use torch.functional.pad to do unequal left side padding, to not compute extra values.

File: pytorch/inference.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    
+    for files in chunker(mel_files, batch_size):
+        mels = []
+        for file_path in files:
+            print(file_path)
+            mel = torch.load(file_path)
+            mel = utils.to_gpu(mel)
+            mels.append(torch.unsqueeze(mel, 0))
+        cond_input = model.get_cond_input(torch.cat(mels, 0))
+        audio_data = wavenet.infer(cond_input, nv_wavenet.Impl.PERSISTENT)
```
Comment (PetrochukM) [inline pytorch/inference.py]: Why use ``PERSISTENT`` instead of ``AUTO``? This should be part of the config.

File: pytorch/inference.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+        for i, file_path in enumerate(files):
+            file_name = os.path.splitext(os.path.basename(file_path))[0]
+            
+            audio = utils.mu_law_decode_numpy(audio_data[i,:].cpu().numpy(), 256)
```
Comment (PetrochukM) [inline pytorch/inference.py]: The Wavenet paper set µ=255:

> where −1 < xt < 1 and µ = 255.

Source: https://arxiv.org/pdf/1609.03499.pdf

File: pytorch/inference.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    """
+    https://stackoverflow.com/a/434328
+    """
+    return (seq[pos:pos + size] for pos in range(0, len(seq), size))
+
+
+def main(mel_files, model_filename, output_dir, batch_size):
+    mel_files = utils.files_to_list(mel_files)
+    model = torch.load(model_filename)['model']
+    wavenet = nv_wavenet.NVWaveNet(**(model.make_nv_wavenet()))
```
Comment (PetrochukM) [inline pytorch/inference.py]: Rename ``make_nv_wavenet``. The function does not make nv wavenet, it exports the WaveNet weights instead.

File: pytorch/mel2samp_onehot.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            # Take segment
+        if audio.size(0) >= self.segment_length:
+            max_audio_start = audio.size(0) - self.segment_length
+            audio_start = random.randint(0, max_audio_start)
+            audio = audio[audio_start:audio_start+self.segment_length]
+        else:
+            audio = torch.nn.functional.pad(audio, (0, self.segment_length - audio.size(0)), 'constant').data
+
+        mel = self.get_mel(audio)
+        audio = utils.mu_law_encode(audio / utils.MAX_WAV_VALUE, self.mu_quantization)
```
Comment (PetrochukM) [inline pytorch/mel2samp_onehot.py]: This does not return a one-hot representation of the audio. This simply returns the audio quantized.

File: pytorch/inference.py (hunk 4 above)
Comment (rafaelvalle) [reply pytorch/inference.py]: https://github.com/RPrenger/nv-wavenet/blob/f3ee1aaa3bc0d0c27863580273e95a1a7b624bde/pytorch/utils.py#L86
Quantization is 256 and µ=255.

File: pytorch/train.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+from mel2samp_onehot import Mel2SampOnehot
+from utils import to_gpu
+
+class CrossEntropyLoss(torch.nn.Module):
+    def __init__(self):
+        super(CrossEntropyLoss, self).__init__()
+        self.num_classes = wavenet_config["n_out_channels"]
+
+    def forward(self, inputs, targets):
+        targets = targets.view(-1)
```
Comment (PetrochukM) [inline pytorch/train.py]: Add a couple comments as towards these random tensor operations.

Comment (rafaelvalle) [reply pytorch/train.py]: Do you think these operations are random or are you just expressing your dissatisfaction in not understanding the code? If you mean the second, let us know what is unclear and we'll make it clearer.

Comment (PetrochukM) [reply pytorch/train.py]: Uhhh. I hope that above comment does not come off as ingenuine. 

I understand that motivation for the above operations because I read the entire above code and replicated WaveNet on my own.

From a locality standpoint of view, the method documentation is not self-sufficient. It's typically bad code style to have functions that do not make sense on their own. 

I understand the code works. I'm happy to drop the issue and move on. It's just a code style thing.

File: pytorch/wavenet.py (hunk 1 above)
Comment (PetrochukM) [reply pytorch/wavenet.py]: https://discuss.pytorch.org/t/can-this-be-achieved/18740/4

File: pytorch/wavenet.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        output = self.conv_end(output)
+
+        # Remove last probabilities because they've seen all the data
+        last = output[:, :, -1]
+        last = last.unsqueeze(2)
+        output = output[:, :, :-1]
+
+        # Replace probability for first value with 0's because we don't know
+        first = last * 0.0
+        output = torch.cat((first, output), dim=2)
```
[the remark this answers is not in the document]
Comment (PetrochukM) [reply pytorch/wavenet.py]: ^ Ditto

[the remark this answers is not in the document]
Comment (RPrenger) [reply pytorch/wavenet.py]: During inference, the first (receptive field) samples are going to be calculated using some padded values, so I don't think it's wrong to include them in the loss function.

File: pytorch/inference.py (hunk 3 above)
Comment (RPrenger) [reply pytorch/inference.py]: Good point!  I've added the change.

File: pytorch/wavenet.py (hunk 1 above)
Comment (RPrenger) [reply pytorch/wavenet.py]: I made it one big 1D conv instead.  I think it trains a little faster now, but should definitely help the latency on inference.

File: pytorch/inference.py (hunk 5 above)
Comment (RPrenger) [reply pytorch/inference.py]: Changed to export_weights

File: pytorch/train.py (hunk 7 above)
Comment (RPrenger) [reply pytorch/train.py]: I've added some comments to make it less opaque.  I'm leaving the names `inputs` and `targets` for now as that's what they use in CrossEntropyLoss

File: pytorch/wavenet.py [hunk 9]
```diff
@@ -123,10 +120,13 @@ def forward(self, forward_input):
        
         forward_input = self.embed(forward_input.long())
         forward_input = forward_input.transpose(1, 2)
-        
+       
+        cond_acts = self.cond_layers(cond_input)
         for i in range(self.n_layers):
             in_act = self.dilate_layers[i](forward_input)
-            in_act = in_act + self.cond_layers[i](cond_input)
+            start = i*2*self.n_residual_channels
```
Comment (PetrochukM) [inline pytorch/wavenet.py]: I think if you do ``cond_acts.view(batch_size, num_layers, block_hidden_size * 2, signal_length)`` you then do not need to compute ``start`` and end``; rather do something like ``cond_acts[:,layer (1, 2, 3..),:,:]``.

File: pytorch/mel2samp_onehot.py (hunk 6 above)
Comment (RPrenger) [reply pytorch/mel2samp_onehot.py]: Quantized audio is the sparse version of the one-hot representation of audio.  I guess it could be confusing if someone wanted to use a dense one-hot representation of audio, but I'm more worried about differentiating this code from code where the sample values are used as actual values, as is often used with Mixture of Logistics sampling.  So I'm going to leave it for now and we can change it if more audio representations come along and it causes confusion (I think we should discourage dense one-hot representations anyway).

File: pytorch/wavenet.py (hunk 2 above)
Comment (RPrenger) [reply pytorch/wavenet.py]: Nice!  And this makes the code shorter too.  I've added the change

File: pytorch/wavenet.py (hunk 9 above)
Comment (RPrenger) [reply pytorch/wavenet.py]: Changed.
